lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -56,14 +56,215 @@ removeNode = (node) ->
 
+updateHeadOrRefresh = (url, upstreamDocument, callback) ->
+  activeAssets = extractTrackedAssets(document)
+  upstreamAssets = extractTrackedAssets(upstreamDocument)
+  updateLinkTags(activeAssets, upstreamAssets)
+  updateScriptTagsOrRefresh(url, activeAssets, upstreamAssets, callback)
+
+extractTrackedAssets = (doc) ->
+  for node in doc.head.childNodes when node.getAttribute?('data-turbolinks-track')?
+    node
+
+insertScriptTask = (scriptNode) ->
+  newNode = document.createElement('SCRIPT')
+  newNode.setAttribute attr.name, attr.value for attr in scriptNode.attributes
+  newNode.setAttribute 'async', false
+  newNode.appendChild document.createTextNode scriptNode.innerHTML
+
+  return (done) ->
+    onScriptEvent = ->
+      newNode.removeEventListener('load', onScriptEvent)
+      newNode.removeEventListener('error', onScriptEvent)
+      done()
+    newNode.addEventListener('load', onScriptEvent)
+    newNode.addEventListener('error', onScriptEvent)
+    document.head.appendChild(newNode)
+    triggerEvent('page:after-script-inserted', newNode)
+
+filterForNodeType = (nodeType) ->
+  (node) -> node.nodeName == nodeType
+
+asyncSeries = (tasks, callback) ->
+  return callback() if tasks.length == 0
+  task = tasks.shift()
+  task(-> asyncSeries(tasks, callback))
+
+updateScriptTagsOrRefresh = (url, activeAssets, upstreamAssets, callback) ->
+  activeScriptTags = activeAssets.filter(filterForNodeType('SCRIPT'))
+  upstreamScriptTags = upstreamAssets.filter(filterForNodeType('SCRIPT'))
+  hasNewSrc = (upstreamNode) ->
+    !(activeScriptTags.some (activeNode) ->
+      upstreamNode.src == activeNode.src)
+  hasExistingScriptAssetName = (upstreamNode) ->
+    activeScriptTags.some (activeNode) ->
+      upstreamNode.attributes['data-turbolinks-track-script-as'].value == activeNode.attributes['data-turbolinks-track-script-as'].value
+
+  scriptsWithNewSrc = upstreamScriptTags.filter(hasNewSrc)
+
+  if scriptsWithNewSrc.some(hasExistingScriptAssetName)
+    return Turbolinks.fullPageNavigate(url.absolute)
+
+  asyncSeries(scriptsWithNewSrc.map(insertScriptTask), callback)
+
+removeStaleLinks = (activeLinks, upstreamLinks) ->
+  remainingActiveLinks = []
+
+  isStaleLink = (link) ->
+    !(upstreamLinks.some (upstreamLink) ->
+      upstreamLink.href == link.href)
+
+  for activeLink, index in activeLinks
+    if isStaleLink(activeLink)
+      removedLink = document.head.removeChild(activeLink)
+      triggerEvent('page:after-link-removed', removedLink)
+    else
+      remainingActiveLinks.push(activeLink)
+
+  remainingActiveLinks
+
+reorderAlreadyExists = (link1, link2, reorders) ->
+  reorders.some (reorderPair) ->
+    link1 in reorderPair && link2 in reorderPair
+
+generateReorderGraph = (activeLinks, upstreamLinks) ->
+  reorders = []
+  for activeLink1 in activeLinks
+    for activeLink2 in activeLinks
+      continue if activeLink1.href == activeLink2.href
+      continue if reorderAlreadyExists(activeLink1, activeLink2, reorders)
```

Comment:
IMO, best to avoid what are basically one line functions used in a single place.